Incorrect GitHub API access when publishing distributions

The following is part of the verbose error message I get when trying to publish a distribution with the latest `dune-release` version `1.01`:

```
dune-release: [EXEC:59307] ['curl' '-s' '-S' '-K' '-' '-D' '-' '--data'
                            '{ "tag_name" : "1.23.0", "body" : "CHANGES:\n\n* Switched to dune and dune-release\n" }'
                            'https://api.github.com/repos/gsl-ocaml/releases/releases']
dune-release: [ERROR] Could not find release id in response:
                      HTTP/1.1 404 Not Found
```

I assume the URL for obtaining release information should read instead as follows:

```
https://api.github.com/repos/mmottl/gsl-ocaml/releases
```

--- a/lib/github.ml
+++ b/lib/github.ml
@@ -268,7 +268,9 @@ let assert_tag_exists ~dry_run tag =
 let publish_distrib ~dry_run ~msg ~archive p =
   let git_for_repo r = Cmd.of_list (Cmd.to_list @@ Vcs.cmd r) in
   let curl = Cmd.(v "curl" % "-L" % "-s" % "-S" % "-K" % "-") in
-  (if dry_run then Ok (D.user, D.repo) else Pkg.distrib_user_and_repo p)
+  (match Pkg.distrib_user_and_repo p with
+  | Error _ as e -> if dry_run then Ok (D.user, D.repo) else e
+  | r -> r)
   >>= fun (user, repo) -> Pkg.tag p
   >>= fun tag ->  assert_tag_exists ~dry_run tag
   >>= fun () -> OS.Cmd.must_exist curl

--- a/lib/pkg.ml
+++ b/lib/pkg.ml
@@ -8,12 +8,13 @@ open Bos_setup
 
 (* Misc *)
 
-let uri_sld uri = match Text.split_uri uri with
-| None -> None
-| Some (_, host, _) ->
-    match List.rev (String.cuts ~sep:"." host) with
-    | _ :: snd :: _ -> Some snd
-    | _ -> None
+let uri_domain uri = match Text.split_uri uri with
+| None -> []
+| Some (_, host, _) -> List.rev (String.cuts ~sep:"." host)
+
+let uri_sld uri = match uri_domain uri with
+| _ :: sld :: _ -> Some sld
+| _ -> None
 
 let uri_append u s = match String.head ~rev:true u with
 | None -> s
@@ -239,8 +240,7 @@ let distrib_uri_of_homepage p =
 
 let distrib_uri ?(raw = false) p =
   let subst_uri p uri =
-    uri
-    >>= fun uri -> name p
+    name p
     >>= fun name -> tag p
     >>= fun tag ->
     let defs = String.Map.(empty |> add "NAME" name |> add "TAG" tag) in
@@ -254,7 +254,15 @@ let distrib_uri ?(raw = false) p =
       | Some u -> Ok u
       | None   -> err_not_found ()
   in
-  if raw then uri else subst_uri p uri
+  uri >>= fun uri ->
+  (match uri_domain uri with
+  | ["io"; "github"; user] ->
+      (match Text.split_uri ~rel:true uri with
+      | None -> R.error_msgf "invalid uri: %s" uri
+      | Some (_, _, path) -> Ok ("https://github.com/" ^ user ^ "/" ^ path))
+  | _ -> Ok uri)
+  >>= fun uri ->
+  if raw then Ok uri else subst_uri p uri
 
 let distrib_filename ?(opam = false) p =
   let sep = if opam then '.' else '-' in

--- a/tests/test_github.ml
+++ b/tests/test_github.ml
@@ -1,28 +1,23 @@
-let user_from_remote_test_case =
-  let test () =
-    let check repo_uri expected =
-      Alcotest.(check (option string))
-        repo_uri
-        expected
-        (Dune_release.Github.user_from_remote repo_uri)
-    in
-    check "git@github.com:username/repo.git" (Some "username");
-    check "git@github.com:user-name/repo.git" (Some "user-name");
-    check "git@github.com:user-name-123/repo.git" (Some "user-name-123");
-    check "git@github.com:123/repo.git" (Some "123");
-    (* Same as above but without the .git part *)
-    check "git@github.com:username/repo" (Some "username");
-    check "git@github.com:user-name/repo" (Some "user-name");
-    check "git@github.com:user-name-123/repo" (Some "user-name-123");
-    check "git@github.com:123/repo" (Some "123");
-    check "wrong" None;
-    check "https://github.com/username/repo.git" None;
-    ()
+let user_from_remote () =
+  let check repo_uri expected =
+    Alcotest.(check (option string))
+      repo_uri
+      expected
+      (Dune_release.Github.user_from_remote repo_uri)
   in
-  ("user_from_remote", `Quick, test)
+  check "git@github.com:username/repo.git" (Some "username");
+  check "git@github.com:user-name/repo.git" (Some "user-name");
+  check "git@github.com:user-name-123/repo.git" (Some "user-name-123");
+  check "git@github.com:123/repo.git" (Some "123");
+  (* Same as above but without the .git part *)
+  check "git@github.com:username/repo" (Some "username");
+  check "git@github.com:user-name/repo" (Some "user-name");
+  check "git@github.com:user-name-123/repo" (Some "user-name-123");
+  check "git@github.com:123/repo" (Some "123");
+  check "wrong" None;
+  check "https://github.com/username/repo.git" None
 
 let suite =
-  ( "Github"
-  , [ user_from_remote_test_case
-    ]
-  )
+  "Github", [
+    "user_from_remote", `Quick, user_from_remote;
+]

--- a/tests/test_tags.ml
+++ b/tests/test_tags.ml
@@ -47,25 +47,27 @@ let distrib_filename () =
   check_true  ~tag:"v0" ~version:"x"  "foo.x"
 
 let distrib_uri () =
-  let check = check ~name:"foo" (fun x -> Pkg.distrib_uri x >>| Fpath.v) in
+  let check = check ~name:"yo" (fun x -> Pkg.distrib_uri x >>| Fpath.v) in
   let dev_repo = ["dev-repo", "git@github.com:foo/bar.git"] in
   let homepage = ["homepage", "https://github.com/foo/bar"] in
   check ~opam:dev_repo ~tag:"v0"
-    "https://github.com/foo/bar/releases/download/v0/foo-v0.tbz";
+    "https://github.com/foo/bar/releases/download/v0/yo-v0.tbz";
   check ~opam:homepage ~tag:"v0"
-    "https://github.com/foo/bar/releases/download/v0/foo-v0.tbz";
+    "https://github.com/foo/bar/releases/download/v0/yo-v0.tbz";
   check ~opam:dev_repo ~version:"v0"
-    "https://github.com/foo/bar/releases/download/v0/foo-v0.tbz";
+    "https://github.com/foo/bar/releases/download/v0/yo-v0.tbz";
   check ~opam:homepage ~version:"v0"
-    "https://github.com/foo/bar/releases/download/v0/foo-v0.tbz";
+    "https://github.com/foo/bar/releases/download/v0/yo-v0.tbz";
   check ~opam:dev_repo ~tag:"v0" ~keep_v:false
-    "https://github.com/foo/bar/releases/download/v0/foo-v0.tbz";
+    "https://github.com/foo/bar/releases/download/v0/yo-v0.tbz";
   check ~opam:homepage ~tag:"v0" ~keep_v:true
-    "https://github.com/foo/bar/releases/download/v0/foo-v0.tbz";
+    "https://github.com/foo/bar/releases/download/v0/yo-v0.tbz";
   check ~opam:dev_repo ~tag:"v0" ~version:"x"
-    "https://github.com/foo/bar/releases/download/v0/foo-v0.tbz";
+    "https://github.com/foo/bar/releases/download/v0/yo-v0.tbz";
   check ~opam:homepage ~tag:"v0" ~version:"x"
-    "https://github.com/foo/bar/releases/download/v0/foo-v0.tbz"
+    "https://github.com/foo/bar/releases/download/v0/yo-v0.tbz";
+  check ~opam:["homepage", "https://foo.github.io/bar"] ~tag:"v0"
+    "https://github.com/foo/bar/releases/download/v0/yo-v0.tbz"
 
 let suite: unit Alcotest.test = "tags", [
   "distrib_file"    , `Quick, distrib_file;

